chainer/testing/function.py

```diff
@@ -0,0 +1,530 @@
+        return grad_outputs
+
+    @_copy_to_test_case()
+    def _generate_grad_grad_inputs(self, inputs_template):
+        grad_grad_inputs = self.generate_grad_grad_inputs(inputs_template)
+        _check_array_types(
+            grad_grad_inputs, _cpu.CpuDevice(), 'generate_grad_grad_inputs')
+        return grad_grad_inputs
+
+    @_copy_to_test_case()
+    def _forward_expected(self, inputs):
+        outputs = self.forward_expected(inputs)
+        _check_array_types(outputs, _cpu.CpuDevice(), 'forward_expected')
+        return outputs
+
+    @_copy_to_test_case()
+    def _forward(self, inputs, backend_config):
+        assert all(isinstance(a, chainer.Variable) for a in inputs)
+        with backend_config:
+            outputs = self.forward(inputs, backend_config.device)
+        _check_variable_types(outputs, backend_config.device, 'forward')
+        if not isinstance(outputs, tuple):
+            outputs = outputs,
+        return outputs
+
+    @_copy_to_test_case(enable_attr_name='forward_test')
+    def test_forward(self, backend_config):
+        """Tests forward computation."""
+
+        self.backend_config = backend_config
+        self.before_test('test_forward')
+
+        cpu_inputs = self._generate_inputs()
+        inputs_copied = [_.copy() for _ in cpu_inputs]
+
+        # Compute expected outputs
+        cpu_expected = self._forward_expected(cpu_inputs)
+        inputs = backend_config.get_array(cpu_inputs)
+        inputs = self._to_noncontiguous_as_needed(inputs)
+
+        # Compute actual outputs
+        outputs = self._forward(
+            tuple(chainer.Variable(a) for a in inputs),
+            backend_config)
+
+        # Check inputs has not changed
+        for i in range(len(inputs)):
+            with FunctionTestError.raise_if_fail(
+                    'Input array has been modified during forward.'):
+                array_module.assert_allclose(
+                    inputs_copied[i], inputs[i], atol=0, rtol=0)
+
+        # Check all outputs are Variable instance
+        for out in outputs:
+            FunctionTestError.check(
+                isinstance(out, chainer.Variable),
+                'Output of forward is not a Variable instance.')
+
+        def format_array_props(msg, expected_arrays, arrays):
+            def format_array(arr):
+                return '{}:{}'.format(arr.shape, arr.dtype.name)
+
+            def format_arrays(arrs):
+                return ', '.join([format_array(a) for a in arrs])
+
+            return (
+                '{}\n'
+                'Expected shapes and dtypes: {}\n'
+                'Actual shapes and dtypes:   {}'.format(
+                    msg,
+                    format_arrays(expected_arrays),
+                    format_arrays(arrays)))
+
+        # Check all outputs are equal to expected values
+        FunctionTestError.check(
+            len(outputs) == len(cpu_expected),
+            'Number of forward outputs is incorrect.')
```

Comment:
How about including the numbers in the message?

Reply:
Fixed.